src/schedule/zephyr_dma_domain.c

```diff
@@ -0,0 +1,298 @@
+#include <sof/schedule/ll_schedule_domain.h>
+#include <sof/schedule/schedule.h>
+#include <sof/schedule/task.h>
+#include <ipc/topology.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+LOG_MODULE_DECLARE(ll_schedule, CONFIG_SOF_LOG_LEVEL);
+
+#if defined(__ZEPHYR__) && defined(CONFIG_IMX)
+#define interrupt_get_irq mux_interrupt_get_irq
+#define interrupt_register mux_interrupt_register
+#define interrupt_unregister mux_interrupt_unregister
+#define interrupt_enable mux_interrupt_enable
+#define interrupt_disable mux_interrupt_disable
+#endif
+
+struct driving_channel {
+	int irq; /* IRQ number */
+	struct dma_chan_data *data; /* channel data */
+	int core; /* core on which the IRQ was enabled */
+	/* pipeline task which registered the DMA IRQ */
+	struct pipeline_task *ptask;
+	bool in_use; /* true if channel is currently in use */
+	void (*work)(void *work_data); /* work to be executed by DMA IRQ */
+	void *work_data; /* data to be used by work function */
+};
+
+struct zephyr_dma_domain {
+	struct dma *dmas; /* array of scheduling DMAs */
+	uint32_t num_dmas; /* number of scheduling DMAs */
+
+	/* driving channels */
+	struct driving_channel channels[PLATFORM_NUM_DMACS][PLATFORM_MAX_DMA_CHAN];
+};
+
+static int zephyr_dma_domain_register(struct ll_schedule_domain *domain,
+				      struct task *task,
+				      void (*work)(void *work_data),
+				      void *work_data);
+static int zephyr_dma_domain_unregister(struct ll_schedule_domain *domain,
+					struct task *task,
+					uint32_t num_tasks);
+static bool zephyr_dma_domain_is_pending(struct ll_schedule_domain *domain,
+					 struct task *task,
+					 struct comp_dev **comp);
+
+const struct ll_schedule_domain_ops zephyr_dma_domain_ops = {
+	.domain_register	= zephyr_dma_domain_register,
+	.domain_unregister	= zephyr_dma_domain_unregister,
+	.domain_is_pending	= zephyr_dma_domain_is_pending
+};
+
+static bool zephyr_dma_domain_is_pending(struct ll_schedule_domain *domain,
+					 struct task *task,
+					 struct comp_dev **comp)
+{
+	/* TODO: actually implement domain_is_pending() */
+	return true;
+}
+
+struct ll_schedule_domain *zephyr_dma_domain_init(struct dma *dmas,
+						  uint32_t num_dmas,
+						  int clk)
+{
+	struct ll_schedule_domain *domain;
+	struct zephyr_dma_domain *zephyr_dma_domain;
+
+	tr_info(&ll_tr, "zepyhr_dma_domain_init(): num_dmas %d, clk %d",
+		num_dmas, clk);
+
+	/* initialize domain */
+	domain = domain_init(SOF_SCHEDULE_LL_DMA, clk,
+			     true, &zephyr_dma_domain_ops);
+
+	/* initialize pdata */
+	zephyr_dma_domain = rzalloc(SOF_MEM_ZONE_SYS_SHARED,
+				    0,
+				    SOF_MEM_CAPS_RAM,
+				    sizeof(*zephyr_dma_domain));
+
+	zephyr_dma_domain->dmas = dmas;
+	zephyr_dma_domain->num_dmas = num_dmas;
+
+	/* set pdata */
+	ll_sch_domain_set_pdata(domain, zephyr_dma_domain);
+
+	return domain;
+}
+
+static void dma_irq_handler(void *data)
+{
+	struct driving_channel *driving = data;
+
+	/* do work */
+	driving->work(driving->work_data);
```

Comment:
I don't think it's a good idea to return to executing the audio (LL) scheduler in interrupt context again. I read your explanation in the description to the PR but I'd still rather keep the scheduler as a Zephyr thread. We want it to be interruptible. I don't know what design considerations you have on iMX, but wouldn't having interruptible per-core scheduling threads be beneficial for you too?

Reply:
Ack - we cant run this in IRQ context. Zephyr is a general purpose RTOS and uses thread priority levels to determine priority for scheduling (xtos used IRQ levels to determine priority).